Content.Client/Actions/ActionsSystem.cs

```diff
@@ -1,302 +1,89 @@
-            }
-
-            UIDirty = true;
-        }
-
-        /// <summary>
-        /// Highlights the item slot (inventory or hand) that contains this item
-        /// </summary>
-        /// <param name="item"></param>
-        public void HighlightItemSlot(EntityUid item)
-        {
-            StopHighlightingItemSlot();
-
-            _highlightedEntity = item;
-            _itemSlotManager.HighlightEntity(item);
-        }
-
-        /// <summary>
-        /// Stops highlighting any item slots we are currently highlighting.
-        /// </summary>H
-        public void StopHighlightingItemSlot()
-        {
-            if (_highlightedEntity == null)
-                return;
-
-            _itemSlotManager.UnHighlightEntity(_highlightedEntity.Value);
-            _highlightedEntity = null;
-        }
-
-        protected override void AddActionInternal(ActionsComponent comp, ActionType action)
-        {
-            // Sometimes the client receives actions from the server, before predicting that newly added components will add
-            // their own shared actions. Just in case those systems ever decided to directly access action properties (e.g.,
-            // action.Toggled), we will remove duplicates:
-            if (comp.Actions.TryGetValue(action, out var existing))
-            {
-                comp.Actions.Remove(existing);
-                Assignments.Replace(existing, action);
-            }
-
-            comp.Actions.Add(action);
-        }
-
-        public override void AddAction(EntityUid uid, ActionType action, EntityUid? provider, ActionsComponent? comp = null, bool dirty = true)
-        {
-            if (!Resolve(uid, ref comp, false))
-                return;
-
-            base.AddAction(uid, action, provider, comp, dirty);
-
-
-            if (uid == _playerManager.LocalPlayer?.ControlledEntity)
-                UIDirty = true;
-        }
-
-        public override void RemoveActions(EntityUid uid, IEnumerable<ActionType> actions, ActionsComponent? comp = null, bool dirty = true)
-        {
-            if (uid != _playerManager.LocalPlayer?.ControlledEntity)
-                return;
-
-            if (!Resolve(uid, ref comp, false))
-                return;
-
-            base.RemoveActions(uid, actions, comp, dirty);
-
-            foreach (var act in actions)
-            {
-                if (act.AutoRemove && !(Ui?.Locked ?? false))
-                    Assignments.Remove(act);
             }
 
-            UIDirty = true;
-        }
-
-        public override void FrameUpdate(float frameTime)
-        {
-            // avoid updating GUI when doing predictions & resetting state.
-            if (UIDirty)
+            foreach (var serverAction in currentState.Actions)
             {
-                UIDirty = false;
-                UpdateUI();
+                if (!component.Actions.TryGetValue(serverAction, out var clientAction))
+                {
+                    added.Add(serverAction);
```

Comment:
This should clone the server action, not add it directly. Otherwise, if the client modifies the action during prediction its actually modifying the server state, which will mess up prediction resetting.

Reply:
Done in [`8b77ff2` (#7202)](https://github.com/space-wizards/space-station-14/pull/7202/commits/8b77ff2f831cc4fca8ce4747a39c0f2c9da58591)